projects/chaos/inc/sequencer_fsm.h

```diff
@@ -57,3 +58,9 @@ void sequencer_fsm_init(void);
 // | Finished state transition |
 // -----------------------------
 StatusCode sequencer_fsm_publish_next_event(const Event *previous_event);
+
+// Delayed variant of event_raise. This version enqueues the highest priority sequence transition it
+// has received as soon as a sequence is completed. This is to allow pre-emption by the Emergency
+// state. Returns true if the submitted event will be raised next. Has knowledge of the present
+// state so will also reject invalid transitions.
+bool sequencer_fsm_event_raise(ChaosEventSequence sequence);
```

Comment:
So this essentially allows you to queue sequences? The naming kinda confuses me.

Why is this necessary? Shouldn't the state transition events have higher priority than the sequenced events? If we hit a fault during sequencing, I think we'd want to immediately enter fault mode and opening everything, not wait for the rest of the events in the sequence.

Reply:
Basically the problem I was trying to fix is that as it was previously Driver Controls could attempt any transition it wanted and it would appear to have succeeded. This is not actually possible since Chaos and Driver Controls could conceivably be in entirely different powers states due to the fact Chaos is able to fault without Driver Controls knowing this. Additionally, the driver could attempt to transition between states faster than Chaos could potentially keep up. To avoid any kind of racing I wanted to ensure that Chaos was always able to tell Driver Controls whether a transition would succeed excluding the case it is interrupted by a fault.

I agree that emergency could be promoted to a higher priority and preempt a transition midway. However, the other states should only occur one at a time.

Consider the case where Driver Controls rapidly requests:
Charge -> Idle -> Drive -> Idle -> Charge -> Idle
It is very likely that under the current scheme we would actually have the following occur
Idle -> Idle -> Idle -> Charge -> Charge -> Drive (fail) So we end up in Charge when we wanted Idle. This is very bad!!!! 
Or due to a race 
Idle -> Idle -> Charge -> Charge -> Idle -> Drive 
The last scenario would be even worse! Presumably this would never occur unless someone is very unlucky but I worry about it happening. I also don't want to switch the event queue to a FIFO because I depend on that behavior for CAN and FSM transition logic but the sequencing itself should be a FIFO. This workaround provides a limited version of that behavior.

Reply:
Hmm, Chaos should never fault without Driver Controls knowing about it since it should be broadcasted over CAN. We need to keep the two synchronized.

If the concern is that the sequence transitions occur out-of-order due to the event queue reordering, it seems like the solution is really to use a FIFO. We can change the event queue to a priority FIFO so your high priority events will take priority. I think the ideal behavior would be for Chaos to override its state as requested by Driver Controls.

Also, the time difference between state changes should be in the order of seconds since we need to hold the power button down.

Reply:
OK so takeaways:
- Chaos should broadcast a fault message when it enters the Emergency state. 
- I agree that a priority FIFO would be the solution. I'll need to wait for that to refactor this PR then.
- I agree that acting on what driver controls says is the ideal behavior other than for faulting. However, I still want to track state on the board since otherwise a bunch of stuff will be in invalid states if somehow an Idle message gets dropped and we try to go from Charge to Drive for example. 

Noted: this was primarily a concern I discovered when thinking about integration tests. In reality I understand that message arrivals will be quite slow.

Reply:
Sounds good. Since the priority FIFO is blocking you you should probably own it. It might make sense to have a few fixed priorities and keep `event_raise` as a default priority.